### PR TITLE
Apply replace-lodash-get codemod to state/reader.

### PR DESCRIPTION
**Pull request by sgomes**
This PR contains automatically generated changes from running the `replace-lodash-get` codemod on `client/state/reader`. I'm planning to do this for each portion of state as it gets modularised.

#### Changes proposed in this Pull Request

* Apply `replace-lodash-get` codemod to `state/reader`.

#### Testing instructions

These are automated changes that should generate equivalent code, with no changes in behaviour. Please double-check that the generated code is sane.

### CODE REVIEW

File: client/state/reader/conversations/selectors/get-reader-conversation-follow-status.js [hunk 1]
```diff
@@ -18,5 +13,5 @@ import 'state/reader/init';
  * @returns {string|null} Conversation follow status (F for following, M for muting, or null)
  */
 export default function getReaderConversationFollowStatus( state, { siteId, postId } ) {
-	return get( state, [ 'reader', 'conversations', 'items', key( siteId, postId ) ], null );
+	return state?.reader?.conversations?.items?.[ key( siteId, postId ) ] ?? null;
```
Comment (tyxla) [inline client/state/reader/conversations/selectors/get-reader-conversation-follow-status.js]: Assuming that `reader` reducer has been loaded, `conversations` and its `items` will always exist.  Also, what is the chance that we'll call the selector with a `null` or `undefined` `state`?

File: client/state/reader/conversations/selectors/is-following-reader-conversation.js [hunk 2]
```diff
@@ -20,7 +15,7 @@ import 'state/reader/init';
  */
 export default function isFollowingReaderConversation( state, { siteId, postId } ) {
 	return (
-		get( state, [ 'reader', 'conversations', 'items', key( siteId, postId ) ] ) ===
+		state?.reader?.conversations?.items?.[ key( siteId, postId ) ] ===
```
Comment (tyxla) [inline client/state/reader/conversations/selectors/is-following-reader-conversation.js]: Assuming that `reader` reducer has been loaded, `conversations` and its `items` will always exist.  Also, what is the chance that we'll call the selector with a `null` or `undefined` `state`?

File: client/state/reader/lists/reducer.js [hunk 3]
```diff
@@ -72,7 +72,7 @@ export const subscribedLists = withSchemaValidation(
 			case READER_LISTS_RECEIVE:
 				return map( action.lists, 'ID' );
 			case READER_LISTS_FOLLOW_SUCCESS:
-				const newListId = get( action, [ 'data', 'list', 'ID' ] );
+				const newListId = action?.data?.list?.ID;
```
Comment (tyxla) [inline client/state/reader/lists/reducer.js]: It's not possible that `action` will be `null` / `undefined` here - if we get to this `switch` case, `action.type` will be equal to `READER_LISTS_FOLLOW_SUCCESS `.

File: client/state/reader/lists/reducer.js [hunk 4]
```diff
@@ -97,7 +97,7 @@ export const subscribedLists = withSchemaValidation(
 export const updatedLists = withSchemaValidation( updatedListsSchema, ( state = [], action ) => {
 	switch ( action.type ) {
 		case READER_LIST_UPDATE_SUCCESS:
-			const newListId = get( action, 'data.list.ID' );
+			const newListId = action?.data?.list?.ID;
```
Comment (tyxla) [inline client/state/reader/lists/reducer.js]: It's not possible that `action` will be `null` / `undefined` here - if we get to this `switch` case, `action.type` will be equal to `READER_LISTS_FOLLOW_SUCCESS `.

File: client/state/reader/posts/reducer.js [hunk 5]
```diff
@@ -26,7 +26,7 @@ export function items( state = {}, action ) {
 	return state;
 }
 export function seen( state = {}, action ) {
-	const id = get( action, 'payload.post.global_ID' );
+	const id = action?.payload?.post?.global_ID;
```
Comment (tyxla) [inline client/state/reader/posts/reducer.js]: What's the chance we'll call the reducer with a `null` / `undefined` action?

File: client/state/reader/posts/selectors.js [hunk 6]
```diff
@@ -47,5 +47,4 @@ export const getPostsByKeys = treeSelect(
 	{ getCacheKey: postKeys => postKeys.map( keyToString ).join() }
 );
 
-export const hasPostBeenSeen = ( state, globalId ) =>
-	!! get( state, [ 'reader', 'posts', 'seen', globalId ] );
+export const hasPostBeenSeen = ( state, globalId ) => !! state?.reader?.posts?.seen?.[ globalId ];
```
Comment (tyxla) [inline client/state/reader/posts/selectors.js]: Assuming that `reader` reducer has been loaded, `posts` and its `seen` will always exist.  Also, what is the chance that we'll call the selector with a `null` or `undefined` `state`?

File: client/state/reader/site-blocks/selectors/get-site-blocks-last-page.js [hunk 7]
```diff
@@ -1,15 +1,10 @@
...
-import { get } from 'lodash';
-
 /**
  * Internal dependencies
  */
 import 'state/reader/init';
 
 export const getSiteBlocksLastPage = state => {
-	return get( state, [ 'reader', 'siteBlocks', 'lastPage' ] );
+	return state?.reader?.siteBlocks?.lastPage;
```
Comment (tyxla) [inline client/state/reader/site-blocks/selectors/get-site-blocks-last-page.js]: Assuming that `reader` reducer has been loaded, `siteBlocks` and its `lastPage` will always exist.  Also, what is the chance that we'll call the selector with a `null` or `undefined` `state`?

File: client/state/reader/site-blocks/selectors/is-fetching-site-blocks.js [hunk 8]
```diff
@@ -13,7 +13,7 @@ import 'state/reader/init';
  * @returns {boolean} true if we are fetching site blocks
  */
 export default function isFetchingSiteBlocks( state ) {
-	const inflightPages = get( state, [ 'reader', 'siteBlocks', 'inflightPages' ] );
+	const inflightPages = state?.reader?.siteBlocks?.inflightPages;
```
Comment (tyxla) [inline client/state/reader/site-blocks/selectors/is-fetching-site-blocks.js]: Assuming that `reader` reducer has been loaded, `siteBlocks` and its `inflightPages` will always exist.  Also, what is the chance that we'll call the selector with a `null` or `undefined` `state`?

File: client/state/reader/tags/images/selectors.js [hunk 9]
```diff
@@ -16,7 +11,7 @@ import 'state/reader/init';
  * @returns {object} Image
  */
 export function getTagImages( state, tag ) {
-	const items = get( state, 'reader.tags.images.items' );
+	const items = state?.reader?.tags?.images?.items;
```
Comment (tyxla) [inline client/state/reader/tags/images/selectors.js]: Assuming that `reader` reducer has been loaded, `tags`, its `images`, and respectively its `items` will always exist.  Also, what is the chance that we'll call the selector with a `null` or `undefined` `state`?

**Review by tyxla — COMMENTED** [review]
Thanks for working on this @sgomes 👍 

I've manually audited all the changes here and left some suggestions.

I am starting to think that the codemod approach isn't a very reliable one - while it does transform the code properly to reflect how it was written before, it doesn't account for whether a piece of state will always be there, thus optional chaining there would be unnecessary.

**Comment by sgomes** [comment]
> I am starting to think that the codemod approach isn't a very reliable one - while it does transform the code properly to reflect how it was written before, it doesn't account for whether a piece of state will always be there, thus optional chaining there would be unnecessary.

The current `_.get` approach effectively does the same, in terms of having optional chaining at all levels of the path. It's just hidden behind the `lodash` machinery.

The codemod is definitely *reliable*, in that it produces correct output, but whether it's useful or not depends on the goal. If the goal is just to get rid of as much `_.get` usage as possible and replace it with standard syntax, then the codemod is doing what we need. If the goal is to audit all of that code and rewrite it in the best possible way then no, the codemod is not sufficient and at best it can only serve as a tool in taking care of the first steps.

I suppose it all depends on which "imperfect" state we prefer: the current usage of `_.get`, or an equally excessively optional version of that, using `?.` instead. I lean towards the latter, but I have a strong bias against `lodash` 🙂

**Comment by tyxla** [comment]
I understand where you're coming from @sgomes. I can also see the value and performance improvement coming from removing all those `get()` instances.

Sorry, I didn't mean to say that the codemod isn't doing a good job. It does a great job, but I have some other considerations for introducing optional chaining everywhere.

I guess, the reason why I'm not extremely happy with this suggestion is that with the lodash usage, it wasn't clear as to what we're expecting to be `null/undefined`. And now, with the optional chaining, we make it explicitly clear what we're expecting to be `null/undefined`, and it doesn't necessarily make sense everywhere, as I've noted in my comments above.

I like what you're suggesting more than `get()` but I'd like us to be more precise and cautious towards where we have optional chaining and not. I just don't want Calypso's codebase in a year from now to have optional chaining everywhere as a synonym for basic `.` chaining 😉

**Comment by jsnajdr** [comment]
> I just don't want Calypso's codebase in a year from now to have optional chaining everywhere as a synonym for basic `.` chaining

Yep, it's already a habit some contributors have: put `a?.b?.c?.d` everywhere by default, no matter what.

**Comment by sgomes** [comment]
Right, I can understand the concern is that the output of the codemod will serve as a set of bad examples that will eventually be followed by developers. But so is the current codebase; leaving `_.get` behind may perpetuate its usage 😕 I guess the only argument for leaving it is that old code may be less preferable as an example than new code, assuming developers check that.

In any case, I'll close this PR, and I'll leave the rest of the codebase similarly unchanged, as unfortunately I don't have the bandwidth to perform a full audit of all of state.

**Pull request closed without merging**